kvserver: Add replica load metrics

kvoli opened the pull request:
Previously, replica stats was oriented for collecting statistics on
batch requests per second - QPS. To facilitate more general use of
additional replica load statistics, this patch introduces more generic
naming in addition to allowing easy opt out of locality specific
elements.

The max, min, count and sum are now recorded per time window as
aggregate, running statistic calculated at write time in constant time.

Previously, writes-per-second representing the count of raft key
mutations applied and queries-per-second, representing the count of
batch requests processed by a replica were tracked on the replica level
by the leaseholder. This patch introduces: requests-per-second, tracking
the number of requests received; reads-per-second, tracking the number
of keys read; write-bytes-per-second and read-bytes-per-second, tracking
the number of bytes that have been read and written to this range per
second.

Only reads-per-second and requests-per-second are exposed in aggregation
over a store as a metric, which is in addition to queries-per-second and
writes-per-second, being currently exported.


Previously, HotRanges would report the replica descriptor and queries
per second (QPS) of the hottest ranges per store; ranked by QPS. This
patch introduces the additional range load information: requests per
second `RQPS`, writes per second `WPS`, reads per second `RPS`, write
bytes per second `WBPS` and read bytes per second `RBPS`. These
additional load metrics are useful for hot range inspection, as they
provide a more granular view of range load, than QPS alone.


resolves https://github.com/cockroachdb/cockroach/issues/72053
resolves https://github.com/cockroachdb/cockroach/issues/77280



Jira issue: CRDB-14773

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76609)
<!-- Reviewable:end -->

kvoli committed:
kvserver: add replica load metrics

Previously, writes-per-second representing the count of raft key
mutations applied and queries-per-second, representing the count of
batch requests processed by a replica were tracked on the replica level
by the leaseholder. This patch introduces: requests-per-second, tracking
the number of requests received; reads-per-second, tracking the number
of keys read; write-bytes-per-second and read-bytes-per-second, tracking
the number of bytes that have been read and written to this range per
second.

Only reads-per-second and requests-per-second are exposed in aggregation
over a store as a metric, which is in addition to queries-per-second and
writes-per-second, being currently exported.

resolves #72053 #77280

Release note (ops change): Added `requests-per-second`, exposed through
the `rebalancing.requestspersecond` metric. `requests-per-second`
tracks the average number of requests received per store, aggregated
over te ranges it contains. Added `reads-per-second`, exposed through
the `relanacing.readspersecond` metric. `reads-per-second` tracks the
count of keys read per second, on a replica basis.

kvoli committed:
kvserver: add replica load info for hotranges

Previously, HotRanges would report the replica descriptor and queries
per second (QPS) of the hottest ranges per store; ranked by QPS. This
patch introduces the additional range load information: requests per
second `RequestsPerSecond`, writes per second `WriteKeysPerSecond`, reads per second `ReadsPerSecond`, write
bytes per second `WriteBytesPerSecond` and read bytes per second `ReadBytesPerSecond`. These
additional load metrics are useful for hot range inspection, as they
provide a more granular view of range load, than QPS alone.

Release note (ops change): HottestRanges will now report additional range
statistics for the reported ranges. These statistics are:
requests per second, the number of requests received by this range
recently per second; writes per second, the number of keys written to in
this range recently per second; reads per second, the number of keys
read from this range recently, per second; write bytes per second, the
number of bytes written to this range recently, per second; and the read
bytes per second, the number of bytes read from this range recently, per
second.

irfansharif reviewed (commented):
(Recording our discussion offline) Given we're not exporting metrics on a per-replica basis, it would be worth seeing what this PR would look like without all of the aggmetric machinery. I suspect it would be simpler -- we could take the same approach we do for the store-level `writes_per_second` metric where we maintain simple integer counters on each replica and have the store semi-periodically iterate through all replicas to update the global count. This latter approach has a better locking profile. By aggmetric-ing everything, every replica stat update linked to the parent store is contending with every other replica on the store. 

We also talked about whether it made sense to track these stats as rates rather than counters. Tracking them as rates/ewma makes reasoning about combining them (during range splits/merges) a little more difficult. If our goal here is to export things to prometheus/http endpoints, tracking them as prom counters at the store levels is sufficient I think. At the replica level where we do may be care about rates, we could do what we're doing for the write QPS today and compute it only on read-time.

abarganier commented:
Great to see more observability on the way! @kvoli can you look into defining some [`AlertingRule`(s)](https://github.com/cockroachdb/cockroach/blob/4cbaaaecd743adada69166216c455168d4345f84/pkg/util/metric/rule.go#L40-L58) and/or [`AggregationRule`(s)](https://github.com/cockroachdb/cockroach/blob/4cbaaaecd743adada69166216c455168d4345f84/pkg/util/metric/rule.go#L60-L70) for this? They can go a long way towards helping people understand expected value ranges/how to effectively use CRDB metrics.

kvoli commented:
> Great to see more observability on the way! @kvoli can you look into defining some [`AlertingRule`(s)](https://github.com/cockroachdb/cockroach/blob/4cbaaaecd743adada69166216c455168d4345f84/pkg/util/metric/rule.go#L40-L58) and/or [`AggregationRule`(s)](https://github.com/cockroachdb/cockroach/blob/4cbaaaecd743adada69166216c455168d4345f84/pkg/util/metric/rule.go#L60-L70) for this? They can go a long way towards helping people understand expected value ranges/how to effectively use CRDB metrics.

Thanks @abarganier, this patch won't actually export per-replica load metrics (due to cardinality it is explicitly disabled). I'm unsure if the store level aggregations that are new (requests-per-second, reads-per-second) are suitable for aggregation/alerting rules.

I'm planning on shaving down this patch quite a bit, see Irfan's comment above. The user visible changes will be:

+ `reads-per-second` `rebalancing.rps`
+ `requests-per-second` `rebalncing.rqps`

Added to exported metrics. Internally, I intend to include the mentioned load metrics per-replica, however these will not be exported to for users (for now, an appropriate aggregation could change that).

kvoli committed:
kvserver: make replica stats operation agnostic

Previously, replica stats was oriented for collecting statistics on
batch requests per second - QPS. To facilitate more general use of
additional replica load statistics, this patch introduces more generic
naming in addition to allowing easy opt out of locality specific
elements.

The max, min, count and sum are now recorded per time window as
aggregate, running staticstic calculated at write time in constant time.

Release note: None

kvoli commented:
Updated @lidorcarmel @irfansharif - I've reworked replica_stats to be more generic for collecting general purpose "stats" about throughput experienced by a replica.

Added in some testing for merges and included write time aggregations.

I was thinking in the future it may be possible to unify the `split` pkg in `kv/kvserver` and `replica_stats.go` in some fashion; most likely using `replica_stats.go` as a base for a splitter.

However that doesn't appear immediately important to this patch. Instead I've gone with the least resistance to get statistics on the throughput signals we discussed: `read_bytes` `read_bytes` `read_keys` and `write_keys` (already tracked). 

This could also deprecate the current `leaseholderStats` and `writeStats` - I'm not sure whether to include immediately or it's worthwhile maintaining duplicate structures for a short period.

Let me know if you have any comments.

lidorcarmel reviewed (changes requested):
> Reviewed 9 of 9 files at r1, 22 of 22 files at r2, 6 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76609)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @kvoli)

___
*[pkg/kv/kvserver/metrics.go, line 299 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76609#-N-UonJo8PqCYfC-ZP0E:-N-UonJo8PqCYfC-ZP0F:b-cm36z) ([raw file](https://github.com/cockroachdb/cockroach/blob/664cc31a207045c810e82799f09d05aed592b2b3/pkg/kv/kvserver/metrics.go#L299)):*
> ```Go
> 		Name:        "rebalancing.requestspersecond",
> 		Help:        "Average number of requests received recently per second.",
> 		Measurement: "Requests",
> ```

should it be `Requests/Sec`?
___
*[pkg/kv/kvserver/replica_metrics.go, line 260 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76609#-N-UuSuG4dOYa40V9-sL:-N-UuSuHDqiNBciECKLi:bm1hwzh) ([raw file](https://github.com/cockroachdb/cockroach/blob/664cc31a207045c810e82799f09d05aed592b2b3/pkg/kv/kvserver/replica_metrics.go#L260)):*
> ```Go
> // RequestsPerSecond returns the range's average requests received per second.
> // A batch request may have one to many requests. Batch requests are
> // represented as queries
> ```

probably can drop the last sentence, but if not then nit: missing period at the end.

also, should the comment say `replica` instead of a `ranges`? same for the comments below.
___
*[pkg/kv/kvserver/replica_stats.go, line 130 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76609#-N-UepNX1eBvygIdtOZ3:-N-UepNX1eBvygIdtOZ4:b-2rhcs1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c23a87d9e32c45a78f09b3193bd3b5ae0cb979c4/pkg/kv/kvserver/replica_stats.go#L130)):*
> ```Go
> // Note that assuming a 50/50 split is optimistic, but it's much better than
> // resetting both sides upon a split.
> // TODO(a-robinson): Write test for this.
> ```

hopefully you added a test, if so please feel free to remove this todo.
___
*[pkg/kv/kvserver/replica_stats.go, line 168 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76609#-N-UkYvk4oR_qa8PiKWd:-N-UkYvk4oR_qa8PiKWe:bb0020i) ([raw file](https://github.com/cockroachdb/cockroach/blob/c23a87d9e32c45a78f09b3193bd3b5ae0cb979c4/pkg/kv/kvserver/replica_stats.go#L168)):*
> ```Go
> 	record.min = math.Min(record.min, count)
> 	record.localityCounts[locality] += count
> 	record.count++
> ```

I don't see where you're using those, and also no tests, I bet I'll see it in the other 2 commits but for future prs it would be nice to have each pr self contained - add a max if it's needed here, and make sure it's tested. for now I think it's fine.
___
*[pkg/kv/kvserver/replica_stats.go, line 115 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76609#-N-Ux9XX3VWzqfAdPV-a:-N-Ux9XX3VWzqfAdPV-b:b-wtncha) ([raw file](https://github.com/cockroachdb/cockroach/blob/664cc31a207045c810e82799f09d05aed592b2b3/pkg/kv/kvserver/replica_stats.go#L115)):*
> ```Go
> 
> 	rsr.sum = rsr.sum / 2.0
> 	other.sum = rsr.sum
> ```

should rsr.count and other.count also be updated here?
___
*[pkg/kv/kvserver/store.go, line 3310 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76609#-N-V31Es2pAjHYG2fs0o:-N-V31Es2pAjHYG2fs0p:b6fbll3) ([raw file](https://github.com/cockroachdb/cockroach/blob/83260add49e8907399ce74d39e8276cbe9ad0ea9/pkg/kv/kvserver/store.go#L3310)):*
> ```Go
> 		hotRepls[i].RPS = topQPS[i].repl.ReadsPerSecond()
> 		hotRepls[i].WBPS = topQPS[i].repl.WriteBytesPerSecond()
> 		hotRepls[i].RBPS = topQPS[i].repl.ReadBytesPerSecond()
> ```

I have a hard time with the short names.. why do we sometimes use long names and sometimes short? can we use the long names in both?
___
*[pkg/server/serverpb/status.proto, line 403 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76609#-N-UzriABMUN6ryjKZzM:-N-UzriABMUN6ryjKZzN:bu3tion) ([raw file](https://github.com/cockroachdb/cockroach/blob/664cc31a207045c810e82799f09d05aed592b2b3/pkg/server/serverpb/status.proto#L403)):*
> ```protobuf
>   //
>   // Note that queries per second will only be known by the leaseholder.
>   // All other replicas will report it as 0.
> ```

is this comment still true? even with follower reads?
___
*[pkg/server/serverpb/status.proto, line 1255 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76609#-N-V3U9O1NgHs1NProsl:-N-V3U9O1NgHs1NProsm:bz4i9ny) ([raw file](https://github.com/cockroachdb/cockroach/blob/83260add49e8907399ce74d39e8276cbe9ad0ea9/pkg/server/serverpb/status.proto#L1255)):*
> ```protobuf
>     ];
> 
>     // Requests per second is the recent number of requests received  per
> ```

nit: extra space.

also - is it worth saying what `recent` means?



_Code quote:_
```protobuf
received  per
```
___
*[pkg/ts/catalog/chart_catalog.go, line 650 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76609#-N-V-80_01g261SXEeyL:-N-V-80_01g261SXEeyM:b-xjufea) ([raw file](https://github.com/cockroachdb/cockroach/blob/664cc31a207045c810e82799f09d05aed592b2b3/pkg/ts/catalog/chart_catalog.go#L650)):*
> ```Go
> 			},
> 			{
> 				Title:   "REQPS",
> ```

the title can be longer I assume? `REQPS` doesn't mean much.. maybe `Requests Per Second`? same for `RPS` below.
___
*[pkg/kv/kvserver/replica_rankings_test.go, line 192 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76609#-N-V-YFRCIMgLqVUPRjC:-N-V-YFRCIMgLqVUPRjD:b-qxxuyi) ([raw file](https://github.com/cockroachdb/cockroach/blob/664cc31a207045c810e82799f09d05aed592b2b3/pkg/kv/kvserver/replica_rankings_test.go#L192)):*
> ```Go
> func genVariableRead(ctx context.Context, start, end roachpb.Key) roachpb.BatchRequest {
> 	scan := roachpb.NewScan(start, end, false)
> 	readBA := roachpb.BatchRequest{}
> ```

nit: should this be `readBa`?



_Code quote:_
```Go
readBA
```
___
*[pkg/kv/kvserver/replica_rankings_test.go, line 237 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76609#-N-V2D8C5uSXXQGzO-kq:-N-V2D8C5uSXXQGzO-kr:bl3tiwe) ([raw file](https://github.com/cockroachdb/cockroach/blob/664cc31a207045c810e82799f09d05aed592b2b3/pkg/kv/kvserver/replica_rankings_test.go#L237)):*
> ```Go
> 		{1, 1, 1, 0, writeSize, 0},
> 		{4, 4, 4, 0, 4 * writeSize, 0},
> 		{64, 64, 64, 0, 64 * writeSize, 0},
> ```

how come you don't test `reads per sec` and `bytes read per sec`?
___
*[pkg/kv/kvserver/replica_rankings_test.go, line 252 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76609#-N-V1pjm79JadoO4vOqz:-N-V1pjm79JadoO4vOr-:bk23zah) ([raw file](https://github.com/cockroachdb/cockroach/blob/664cc31a207045c810e82799f09d05aed592b2b3/pkg/kv/kvserver/replica_rankings_test.go#L252)):*
> ```Go
> 
> 	for _, testCase := range testCases {
> 		// This test can flake, where an errant request - not sent here
> ```

probably worth verifying flakiness if you haven't already, by running bazel with --runs_per_test=1000 or something like that, and --test_filter= this test only.
___
*[pkg/kv/kvserver/replica_stats_test.go, line 399 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76609#-N-Ul-rgA5jsaau5ZuVo:-N-Ul-rgA5jsaau5ZuVp:b-bj42s1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c23a87d9e32c45a78f09b3193bd3b5ae0cb979c4/pkg/kv/kvserver/replica_stats_test.go#L399)):*
> ```Go
> 			rs.recordCount(float64(i*1*multiplier), 1)
> 			rs.recordCount(float64(i*2*multiplier), 2)
> 			rs.recordCount(float64(i*3*multiplier), 3)
> ```

where are the 1,2,3 coming from? also, probably makes sense to have a comment for this function?
___
*[pkg/kv/kvserver/replica_stats_test.go, line 512 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76609#-N-UmETr-sppOHyjEDkg:-N-UmETr-sppOHyjEDkh:b7f2r0) ([raw file](https://github.com/cockroachdb/cockroach/blob/c23a87d9e32c45a78f09b3193bd3b5ae0cb979c4/pkg/kv/kvserver/replica_stats_test.go#L512)):*
> ```Go
> }
> 
> func genTestingReplicaStats(windowedMultipliers [6]int, n, offset int) *replicaStats {
> ```

looks like this is the same as above?

and I see now it is removed in this commit (3), should have been removed in commit 1?


<!-- Sent from Reviewable.io -->

kvoli reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76609)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @irfansharif, and @lidorcarmel)

___
*[pkg/kv/kvserver/metrics.go, line 299 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76609#-N-UonJo8PqCYfC-ZP0E:-N-Zicli8-B_iddRqxY9:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/664cc31a207045c810e82799f09d05aed592b2b3/pkg/kv/kvserver/metrics.go#L299)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

should it be `Requests/Sec`?
</blockquote></details>

Updated.
___
*[pkg/kv/kvserver/replica_metrics.go, line 260 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76609#-N-UuSuG4dOYa40V9-sL:-N-ZifWjAJDHCMiA0vrP:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/664cc31a207045c810e82799f09d05aed592b2b3/pkg/kv/kvserver/replica_metrics.go#L260)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

probably can drop the last sentence, but if not then nit: missing period at the end.

also, should the comment say `replica` instead of a `ranges`? same for the comments below.
</blockquote></details>

Updated.
___
*[pkg/kv/kvserver/replica_stats.go, line 130 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76609#-N-UepNX1eBvygIdtOZ3:-N-Zilnw9ErxVuP1GiNE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c23a87d9e32c45a78f09b3193bd3b5ae0cb979c4/pkg/kv/kvserver/replica_stats.go#L130)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

hopefully you added a test, if so please feel free to remove this todo.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/replica_stats.go, line 168 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76609#-N-UkYvk4oR_qa8PiKWd:-N-ZinNE9fh_HO8zIuF6:b2nhtan) ([raw file](https://github.com/cockroachdb/cockroach/blob/c23a87d9e32c45a78f09b3193bd3b5ae0cb979c4/pkg/kv/kvserver/replica_stats.go#L168)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

I don't see where you're using those, and also no tests, I bet I'll see it in the other 2 commits but for future prs it would be nice to have each pr self contained - add a max if it's needed here, and make sure it's tested. for now I think it's fine.
</blockquote></details>

Yup
___
*[pkg/kv/kvserver/replica_stats.go, line 115 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76609#-N-Ux9XX3VWzqfAdPV-a:-N-ZihJo-3WY4-Ol0IvI:b70jdye) ([raw file](https://github.com/cockroachdb/cockroach/blob/664cc31a207045c810e82799f09d05aed592b2b3/pkg/kv/kvserver/replica_stats.go#L115)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

should rsr.count and other.count also be updated here?
</blockquote></details>

Updated they should. Added a test here as well.
___
*[pkg/kv/kvserver/store.go, line 3310 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76609#-N-V31Es2pAjHYG2fs0o:-N-ZisVVEwa5X5sFy7GW:b-3s8lz4) ([raw file](https://github.com/cockroachdb/cockroach/blob/83260add49e8907399ce74d39e8276cbe9ad0ea9/pkg/kv/kvserver/store.go#L3310)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

I have a hard time with the short names.. why do we sometimes use long names and sometimes short? can we use the long names in both?
</blockquote></details>

Renamed to be more verbose.
___
*[pkg/server/serverpb/status.proto, line 403 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76609#-N-UzriABMUN6ryjKZzM:-N-ZivcNEIT9HOaYDdSg:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/664cc31a207045c810e82799f09d05aed592b2b3/pkg/server/serverpb/status.proto#L403)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

is this comment still true? even with follower reads?
</blockquote></details>

Updated.
___
*[pkg/server/serverpb/status.proto, line 1255 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76609#-N-V3U9O1NgHs1NProsl:-N-ZiwnpDRX3iOgZN1IG:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/83260add49e8907399ce74d39e8276cbe9ad0ea9/pkg/server/serverpb/status.proto#L1255)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

nit: extra space.

also - is it worth saying what `recent` means?
</blockquote></details>

Updated.
___
*[pkg/ts/catalog/chart_catalog.go, line 650 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76609#-N-V-80_01g261SXEeyL:-N-ZixkSC_7hMocztVFy:b-v1jz6z) ([raw file](https://github.com/cockroachdb/cockroach/blob/664cc31a207045c810e82799f09d05aed592b2b3/pkg/ts/catalog/chart_catalog.go#L650)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

the title can be longer I assume? `REQPS` doesn't mean much.. maybe `Requests Per Second`? same for `RPS` below.
</blockquote></details>

Updated to be more verbose.
___
*[pkg/kv/kvserver/replica_rankings_test.go, line 192 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76609#-N-V-YFRCIMgLqVUPRjC:-N-Zj0ABB_WqsweNA8W3:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/664cc31a207045c810e82799f09d05aed592b2b3/pkg/kv/kvserver/replica_rankings_test.go#L192)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

nit: should this be `readBa`?
</blockquote></details>

Updated.
___
*[pkg/kv/kvserver/replica_rankings_test.go, line 237 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76609#-N-V2D8C5uSXXQGzO-kq:-N-Zj15D8lZ6ZVrNfw9M:b-l0c3nc) ([raw file](https://github.com/cockroachdb/cockroach/blob/664cc31a207045c810e82799f09d05aed592b2b3/pkg/kv/kvserver/replica_rankings_test.go#L237)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

how come you don't test `reads per sec` and `bytes read per sec`?
</blockquote></details>

They are tested in below.
___
*[pkg/kv/kvserver/replica_rankings_test.go, line 252 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76609#-N-V1pjm79JadoO4vOqz:-N-Zj37b1ZicP7pivLUY:bkgwxoj) ([raw file](https://github.com/cockroachdb/cockroach/blob/664cc31a207045c810e82799f09d05aed592b2b3/pkg/kv/kvserver/replica_rankings_test.go#L252)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

probably worth verifying flakiness if you haven't already, by running bazel with --runs_per_test=1000 or something like that, and --test_filter= this test only.
</blockquote></details>

I've tested for about 10k runs, will still investigate other solutions here but its not too flaky.
___
*[pkg/kv/kvserver/replica_stats_test.go, line 399 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76609#-N-Ul-rgA5jsaau5ZuVo:-N-ZjC3HAJwZa0iwyKcW:bs3mhdg) ([raw file](https://github.com/cockroachdb/cockroach/blob/c23a87d9e32c45a78f09b3193bd3b5ae0cb979c4/pkg/kv/kvserver/replica_stats_test.go#L399)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

where are the 1,2,3 coming from? also, probably makes sense to have a comment for this function?
</blockquote></details>

Comment added.
___
*[pkg/kv/kvserver/replica_stats_test.go, line 512 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76609#-N-UmETr-sppOHyjEDkg:-N-ZjEYZ5L2Dnow3_bTf:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/c23a87d9e32c45a78f09b3193bd3b5ae0cb979c4/pkg/kv/kvserver/replica_stats_test.go#L512)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

looks like this is the same as above?

and I see now it is removed in this commit (3), should have been removed in commit 1?
</blockquote></details>

Fixed.


<!-- Sent from Reviewable.io -->

lidorcarmel reviewed (approved):
> Reviewed 36 of 50 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76609)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @lidorcarmel)



<!-- Sent from Reviewable.io -->

irfansharif commented on pkg/kv/kvserver/replica_load.go:
```diff
@@ -0,0 +1,66 @@
+
+// ReplicaLoad tracks a sliding window of throughput on a replica. By default,
+// there are 6, 5 minute sliding windows.
+type ReplicaLoad struct {
```
Do we want a mutex at this level instead of the individual replicaStats levels?

kvoli replied:
I don't think so given they will be called into, within different 3 separate parts of a request/response lifecycle so finer grained locking seems more appropriate here.

irfansharif replied:
Are those three parts contending on access here? Especially non of these locks are long-held? Code like the following feels a lot more cumbersome relative if we had a single lock:

```go
			repl.loadStats.requests.mu.Unlock()
 			repl.loadStats.writeKeys.mu.Unlock()
 			repl.loadStats.readKeys.mu.Unlock()
 			repl.loadStats.writeBytes.mu.Unlock()
 			repl.loadStats.readBytes.mu.Unlock()
 			repl.loadStats.batchRequests.mu.Unlock()
```

It's not clear to me that partitioning the locks here by field of access is giving us measurable performance benefit, so I'd err on the side of doing the simpler thing until needing to otherwise (demonstrated by a benchmark).

kvoli replied:
The other reason was to not deprecate the existing qps recording struct leaseholderStats  as of yet. I agree that it might not have any noticeable impact on latency.

irfansharif commented on pkg/server/serverpb/status.proto:
```diff
@@ -399,11 +399,18 @@ message RangeProblems {
   // Writes per second served by this range.
   double writes_per_second = 2;
+  // Requests per second served by this range.
+  double requests_per_second = 3;
```
A bit of commentary talking about the relation between requests_per_second, writes_per_second and reads_per_seconds would help (even if trivial).

kvoli replied:
Good idea. Updated.

irfansharif reviewed (approved):
Took only the most cursory pass but this much nicer than we had previously!

kvoli commented:
bors r+

tbg commented:
Late to the party, sorry!

Re:

Only reads-per-second and requests-per-second are exposed in aggregation
over a store as a metric

Why?

kvoli commented:
> Late to the party, sorry!
> 
> Re:
> 
> Only reads-per-second and requests-per-second are exposed in aggregation
> over a store as a metric
> 
> Why?

All the others added at a replica level are already exported. So this are the missing two. Batch reqs qps, write keys as writes per second, disk throughput is tracked on hardware (we may want to export the req potion of it however).

For allocator changes that take into account finer grained stats, these aggregates would also need to be gosipped.
